server/services/v1/billing/reporter.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2022-2023 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package billing
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/rs/zerolog/log"
+	"github.com/tigrisdata/tigris/errors"
+	"github.com/tigrisdata/tigris/server/config"
+	"github.com/tigrisdata/tigris/server/metadata"
+	"github.com/tigrisdata/tigris/server/metrics"
+	ulog "github.com/tigrisdata/tigris/util/log"
+)
+
+type UsageReporter struct {
+	glbState   metrics.UsageProvider
+	billingSvc Provider
+	interval   time.Duration
+	tenantMgr  metadata.NamespaceMetadataMgr
+	ctx        context.Context
+}
+
+func NewUsageReporter(gs metrics.UsageProvider, tm metadata.NamespaceMetadataMgr, billing Provider) (*UsageReporter, error) {
+	if gs == nil || tm == nil {
+		return nil, errors.Internal("usage reporter cannot be initialized")
+	}
+
+	return &UsageReporter{
+		glbState:   gs,
```

Comment:
For the global status, there should be only one existing per server instance. Because there is only a global one, there is no need to pass it. Do we have a good way to handle server level singletons like that? This was one of the areas that I didn't like in the global status code. Metric providers have a similar problem.

Reply:
I don't think there is a framework as such that can do that. The `NewGlobalStatus()` method can do a nil check to only initialize `var GlobalSt *GlobalStatus` once.

Reply:
`sync.Once` can be used to guarantee that it is only initialized once https://pkg.go.dev/sync#Once